File: CS/CustomTsaClient/BouncyCastleTsaClient.cs
using DevExpress.Office.Tsp;
using Org.BouncyCastle.Asn1.Cmp;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Security;
using Org.BouncyCastle.Tsp;
using System;
using System.IO;
using System.Net.Http;
using System.Security.Cryptography;

public class BouncyCastleTsaClient : ITsaClient
{
    readonly Uri tsaServerURI;

    readonly IDigest hashCalculator;

    readonly HttpClient httpClient;

    public BouncyCastleTsaClient(Uri tsaServerURI, IDigest hashCalculator, HttpClient httpClient)
    {
        this.tsaServerURI = tsaServerURI;
        this.hashCalculator = hashCalculator;
        this.httpClient = httpClient;
    }
    byte[] CalculateDigest(Stream stream)
    {
        byte[] buffer = new byte[81920];
        hashCalculator.Reset();
        int bytesRead;
        while ((bytesRead = stream.Read(buffer, 0, buffer.Length)) > 0)
            hashCalculator.BlockUpdate(buffer, 0, bytesRead);
        byte[] result = new byte[hashCalculator.GetDigestSize()];
        hashCalculator.DoFinal(result, 0);
        return result;
    }

    public byte[] GenerateTimeStamp(Stream stream)
    {
        //Generate a timestamp request:
        TimeStampRequestGenerator tsqGenerator = new TimeStampRequestGenerator();
        tsqGenerator.SetCertReq(true);
        BigInteger nonce;
        using (RandomNumberGenerator generator = RandomNumberGenerator.Create())
        {
            byte[] nonceValue = new byte[10];
            generator.GetBytes(nonceValue);
            nonce = new BigInteger(nonceValue);
        }
        string algorithmOid = DigestUtilities.GetObjectIdentifier(hashCalculator.AlgorithmName).Id;
        TimeStampRequest request = tsqGenerator.Generate(algorithmOid, CalculateDigest(stream), nonce);
        byte[] requestBytes = request.GetEncoded();

        //Send the request to a server:
        HttpClient httpClient = new HttpClient();
        using var content = new ByteArrayContent(requestBytes);
        content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/timestamp-query");
        //Get a responce from the server:
        using HttpResponseMessage responseMessage = httpClient.PostAsync(tsaServerURI, content).Result;
        if (!responseMessage.IsSuccessStatusCode)
        {
            throw new Exception($"TimeStamp request to the \"{tsaServerURI}\" failed with status code {responseMessage.StatusCode}.");
        }
        byte[] responseBytes = responseMessage.Content.ReadAsByteArrayAsync().Result;
        //Read the response:
        TimeStampResponse response = new TimeStampResponse(responseBytes);
        response.Validate(request);
        PkiFailureInfo failure = response.GetFailInfo();

        //Throw an exception if the responce returned an error:                
        if (failure != null)
            throw new Exception($"TimeStamp request to the \"{tsaServerURI}\" failed.");
        TimeStampToken token = response.TimeStampToken;

        //Throw an exception if the responce doesn't contain the timestamp:
        if (token == null)
            throw new Exception($"TimeStamp request to the \"{tsaServerURI}\" failed.");
        return token.GetEncoded();
    }

    public byte[] GenerateTimeStamp(byte[] digest, string digestAlgorithmOID)
    {
        //Generate a timestamp request:
        TimeStampRequestGenerator tsqGenerator = new TimeStampRequestGenerator();
        tsqGenerator.SetCertReq(true);
        BigInteger nonce;
        using (RandomNumberGenerator generator = RandomNumberGenerator.Create())
        {
            byte[] nonceValue = new byte[10];
            generator.GetBytes(nonceValue);
            nonce = new BigInteger(nonceValue);
        }
        string algorithmOid = DigestUtilities.GetObjectIdentifier(hashCalculator.AlgorithmName).Id;
        TimeStampRequest request = tsqGenerator.Generate(algorithmOid, digest, nonce);
        byte[] requestBytes = request.GetEncoded();

        //Send the request to a server:
        HttpClient httpClient = new HttpClient();
        using var content = new ByteArrayContent(requestBytes);
        content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/timestamp-query");
        //Get a responce from the server:
        using HttpResponseMessage responseMessage = httpClient.PostAsync(tsaServerURI, content).Result;
        if (!responseMessage.IsSuccessStatusCode)
        {
            throw new Exception($"TimeStamp request to the \"{tsaServerURI}\" failed with status code {responseMessage.StatusCode}.");
        }
        byte[] responseBytes = responseMessage.Content.ReadAsByteArrayAsync().Result;

        //Read the response:
        TimeStampResponse response = new TimeStampResponse(responseBytes);
        response.Validate(request);
        PkiFailureInfo failure = response.GetFailInfo();


        //Throw an exception if the responce returned an error:                
        if (failure != null)
            throw new Exception($"TimeStamp request to the \"{tsaServerURI}\" failed.");
        TimeStampToken token = response.TimeStampToken;

        //Throw an exception if the responce doesn't contain the timestamp:
        if (token == null)
            throw new Exception($"TimeStamp request to the \"{tsaServerURI}\" failed.");
        return token.GetEncoded();
    }
}

File: VB/CustomTsaClient/BouncyCastleTsaClient.vb
Imports DevExpress.Office.Tsp
Imports Org.BouncyCastle.Asn1.Cmp
Imports Org.BouncyCastle.Crypto
Imports Org.BouncyCastle.Math
Imports Org.BouncyCastle.Security
Imports Org.BouncyCastle.Tsp
Imports System
Imports System.IO
Imports System.Net.Http
Imports System.Security.Cryptography

Public Class BouncyCastleTsaClient
    Implements ITsaClient

    Private ReadOnly tsaServerURI As Uri
    Private ReadOnly hashCalculator As IDigest
    Private ReadOnly httpClient As HttpClient

    Public Sub New(tsaServerURI As Uri, hashCalculator As IDigest, httpClient As HttpClient)
        Me.tsaServerURI = tsaServerURI
        Me.hashCalculator = hashCalculator
        Me.httpClient = httpClient
    End Sub

    Private Function CalculateDigest(stream As Stream) As Byte()
        Dim buffer(81919) As Byte
        hashCalculator.Reset()
        Dim bytesRead As Integer
        Do
            bytesRead = stream.Read(buffer, 0, buffer.Length)
            If bytesRead > 0 Then
                hashCalculator.BlockUpdate(buffer, 0, bytesRead)
            End If
        Loop While bytesRead > 0
        Dim result(hashCalculator.GetDigestSize() - 1) As Byte
        hashCalculator.DoFinal(result, 0)
        Return result
    End Function

    Public Function GenerateTimeStamp(stream As Stream) As Byte() Implements ITsaClient.GenerateTimeStamp
        Dim tsqGenerator As New TimeStampRequestGenerator()
        tsqGenerator.SetCertReq(True)
        Dim nonce As BigInteger
        Using generator As RandomNumberGenerator = RandomNumberGenerator.Create()
            Dim nonceValue(9) As Byte
            generator.GetBytes(nonceValue)
            nonce = New BigInteger(nonceValue)
        End Using
        Dim algorithmOid As String = DigestUtilities.GetObjectIdentifier(hashCalculator.AlgorithmName).Id
        Dim request As TimeStampRequest = tsqGenerator.Generate(algorithmOid, CalculateDigest(stream), nonce)
        Dim requestBytes As Byte() = request.GetEncoded()

        Using content As New ByteArrayContent(requestBytes)
            content.Headers.ContentType = New System.Net.Http.Headers.MediaTypeHeaderValue("application/timestamp-query")
            Using responseMessage As HttpResponseMessage = httpClient.PostAsync(tsaServerURI, content).Result
                If Not responseMessage.IsSuccessStatusCode Then
                    Throw New Exception($"TimeStamp request to ""{tsaServerURI}"" failed with status code {responseMessage.StatusCode}.")
                End If
                Dim responseBytes As Byte() = responseMessage.Content.ReadAsByteArrayAsync().Result
                Dim response As New TimeStampResponse(responseBytes)
                response.Validate(request)
                Dim failure As PkiFailureInfo = response.GetFailInfo()
                If failure IsNot Nothing Then
                    Throw New Exception($"TimeStamp request to the ""{tsaServerURI}"" failed.")
                End If
                Dim token As TimeStampToken = response.TimeStampToken
                If token Is Nothing Then
                    Throw New Exception($"TimeStamp request to ""{tsaServerURI}"" failed.")
                End If
                Return token.GetEncoded()
            End Using
        End Using
    End Function

    Public Function GenerateTimeStamp(digest As Byte(), digestAlgorithmOID As String) As Byte() Implements ITsaClient.GenerateTimeStamp
        Dim tsqGenerator As New TimeStampRequestGenerator()
        tsqGenerator.SetCertReq(True)
        Dim nonce As BigInteger
        Using generator As RandomNumberGenerator = RandomNumberGenerator.Create()
            Dim nonceValue(9) As Byte
            generator.GetBytes(nonceValue)
            nonce = New BigInteger(nonceValue)
        End Using
        Dim algorithmOid As String = DigestUtilities.GetObjectIdentifier(hashCalculator.AlgorithmName).Id
        Dim request As TimeStampRequest = tsqGenerator.Generate(algorithmOid, digest, nonce)
        Dim requestBytes As Byte() = request.GetEncoded()

        Using content As New ByteArrayContent(requestBytes)
            content.Headers.ContentType = New System.Net.Http.Headers.MediaTypeHeaderValue("application/timestamp-query")
            Using responseMessage As HttpResponseMessage = httpClient.PostAsync(tsaServerURI, content).Result
                If Not responseMessage.IsSuccessStatusCode Then
                    Throw New Exception($"TimeStamp request to ""{tsaServerURI}"" failed with status code {responseMessage.StatusCode}.")
                End If
                Dim responseBytes As Byte() = responseMessage.Content.ReadAsByteArrayAsync().Result
                Dim response As New TimeStampResponse(responseBytes)
                response.Validate(request)
                Dim failure As PkiFailureInfo = response.GetFailInfo()
                If failure IsNot Nothing Then
                    Throw New Exception($"TimeStamp request to ""{tsaServerURI}"" failed.")
                End If
                Dim token As TimeStampToken = response.TimeStampToken
                If token Is Nothing Then
                    Throw New Exception($"TimeStamp request to ""{tsaServerURI}"" failed.")
                End If
                Return token.GetEncoded()
            End Using
        End Using
    End Function

End Class
